training/ikon20/part2-working-with-masks.ipynb

```diff
@@ -0,0 +1,256 @@
+    "   - If you run into an error regarding a length mismatch when inserting the coordinate, remember that `'tof'` is a bin-edge coordinate, i.e., it is by 1 longer than the number of bins.\n",
+    "     Use, e.g., only the left bin edges, i.e., all but the last, to create the masks.\n",
+    "2. Use the HTML view and plot the data after masking to explore the effect.\n",
+    "3. Pass a `dict` containing `counts` (computed above as `counts = sc.sum(data, 'tof')`) and the equivalent counts computed *after* masking to `sc.plot.plot`.\n",
+    "   Use this to verify that the promt-pulse mask results in removal of counts."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tof = data.coords['tof']\n",
+    "data.masks['promt-pulse'] = (tof['tof',:-1] > 17500.0 * sc.units.us) & (tof['tof',1:] < 19000.0 * sc.units.us)\n",
+    "sc.plot.plot({'before':counts, 'after':sc.sum(data, 'tof')})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercise 2: Masking spatially\n",
+    "\n",
+    "By masking an `x` range, mask the end of the tubes.\n",
+    "- Define `x = sc.geometry.x(data.coords['position'])` to extract only the x-component of the position coordinate.\n",
+    "- Create the masks.\n",
+    "- Use the instrument view (`sc.neutron.instrumentview(data)`) to inspect the result."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "x = sc.geometry.x(data.coords['position'])\n",
+    "data.masks['tube-ends'] = x < -0.2 * sc.units.m\n",
+    "sc.neutron.instrument_view(data)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercise 3: Combining conditions\n",
+    "\n",
+    "Mask the broken pixels with zero counts near the beam stop (center).\n",
+    "- Note that there are pixels at larger scattering angles (larger x) which have real zeros.\n",
+    "  These should not be masked.\n",
+    "- Combine the condition for zero counts with a spatial mask, e.g., based on `x`, to ensure the masks takes only effect close to the direct beam / beam stop."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# This would mask too much, what needs to be added?\n",
+    "counts.data == 0.0 * sc.units.counts"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "broken = (counts.data == 0.0 * sc.units.counts) & (sc.abs(x) < 0.1 * sc.units.m)\n",
+    "data.masks['bad-pixels'] = broken\n",
+    "sc.neutron.instrument_view(data, pixel_size=0.01)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercise 4: More spatial masking\n",
+    "\n",
+    "Pick one (or more, if desired):\n",
```

Comment:
I think it is really good (and very important) that we start to look at and mask the data in terms of theta angles (i.e. physical quantities and not spectrum ids). It is also very good to show the plot with theta at the end. :+1: 

Maybe it's worth emphasising that you can work with your data in theta early on, and forget about the spectrum dimension if you wish so? (just like you can just work in `x, y, tof` in imaging)

Reply:
Not sure, you still need to do all the detector corrections, possibly masking, and other background subtractions while in detector space. So "early on" may actually be quite late?